--- a/hibernate-core/src/main/java/org/hibernate/EmptyInterceptor.java
+++ b/hibernate-core/src/main/java/org/hibernate/EmptyInterceptor.java
@@ -111,10 +111,10 @@ public class EmptyInterceptor implements Interceptor, Serializable {
 		return sql;
 	}
 
-	public void onCollectionRemove(Object collection, Serializable key) throws CallbackException {}
+	public boolean onCollectionRemove(Object collection, Serializable key) throws CallbackException { return false; }
 
-	public void onCollectionRecreate(Object collection, Serializable key) throws CallbackException {}
+	public boolean onCollectionRecreate(Object collection, Serializable key) throws CallbackException { return false; }
 
-	public void onCollectionUpdate(Object collection, Serializable key) throws CallbackException {}
+	public boolean onCollectionUpdate(Object collection, Serializable key) throws CallbackException { return false; }
 	
-}
\ No newline at end of file
+}
--- a/hibernate-core/src/main/java/org/hibernate/Interceptor.java
+++ b/hibernate-core/src/main/java/org/hibernate/Interceptor.java
@@ -132,7 +132,7 @@ public interface Interceptor {
 	 *
 	 * @throws CallbackException Thrown if the interceptor encounters any problems handling the callback.
 	 */
-	public void onCollectionRecreate(Object collection, Serializable key) throws CallbackException;
+	public boolean onCollectionRecreate(Object collection, Serializable key) throws CallbackException;
 
 	/**
 	 * Called before a collection is deleted.
@@ -142,7 +142,7 @@ public interface Interceptor {
 	 *
 	 * @throws CallbackException Thrown if the interceptor encounters any problems handling the callback.
 	 */
-	public void onCollectionRemove(Object collection, Serializable key) throws CallbackException;
+	public boolean onCollectionRemove(Object collection, Serializable key) throws CallbackException;
 
 	/**
 	 * Called before a collection is updated.
@@ -152,7 +152,7 @@ public interface Interceptor {
 	 *
 	 * @throws CallbackException Thrown if the interceptor encounters any problems handling the callback.
 	 */
-	public void onCollectionUpdate(Object collection, Serializable key) throws CallbackException;
+	public boolean onCollectionUpdate(Object collection, Serializable key) throws CallbackException;
 
 	/**
 	 * Called before a flush
--- a/hibernate-core/src/main/java/org/hibernate/event/internal/AbstractFlushingEventListener.java
+++ b/hibernate-core/src/main/java/org/hibernate/event/internal/AbstractFlushingEventListener.java
@@ -147,6 +147,27 @@ public abstract class AbstractFlushingEventListener implements Serializable {
 		for ( Map.Entry me : IdentityMap.concurrentEntries( persistenceContext.getEntityEntries() ) ) {
 			EntityEntry entry = (EntityEntry) me.getValue();
 			Status status = entry.getStatus();
+			
+             		// This entity will be saved?
+                       boolean willBeSaved = true;
+                        try {
+                               Object o = me.getKey();
+                               Class c = o.getClass();
+                               Class jpaBase = Class.forName("play.db.jpa.JPABase");
+                               while(!c.equals(Object.class)) {
+                                       if(c.equals(jpaBase)) {
+                                               willBeSaved = (Boolean)(jpaBase.getDeclaredField("willBeSaved").get(o));
+                                               break;
+                                       }
+                                       c = c.getSuperclass();
+                               }
+                               if(!willBeSaved) {
+                                       continue;
+                               }
+                       } catch(Exception e) {
+                               e.printStackTrace();
+                       }
+
 			if ( status == Status.MANAGED || status == Status.SAVING || status == Status.READ_ONLY ) {
 				cascadeOnFlush( session, entry.getPersister(), me.getKey(), anything );
 			}
@@ -257,31 +278,35 @@ public abstract class AbstractFlushingEventListener implements Serializable {
 			CollectionEntry ce = me.getValue();
 
 			if ( ce.isDorecreate() ) {
-				session.getInterceptor().onCollectionRecreate( coll, ce.getCurrentKey() );
-				actionQueue.addAction(
-						new CollectionRecreateAction(
-								coll,
-								ce.getCurrentPersister(),
-								ce.getCurrentKey(),
-								session
-							)
-					);
+				if ( session.getInterceptor().onCollectionRecreate( coll, ce.getCurrentKey() ) ) {
+                                       actionQueue.addAction(
+                                                       new CollectionRecreateAction(
+                                                                       coll,
+                                                                       ce.getCurrentPersister(),
+                                                                       ce.getCurrentKey(),
+                                                                       session
+                                                               )
+                                               );
+                               }
+
 			}
 			if ( ce.isDoremove() ) {
-				session.getInterceptor().onCollectionRemove( coll, ce.getLoadedKey() );
-				actionQueue.addAction(
-						new CollectionRemoveAction(
-								coll,
-								ce.getLoadedPersister(),
-								ce.getLoadedKey(),
-								ce.isSnapshotEmpty(coll),
-								session
-							)
-					);
+				if ( session.getInterceptor().onCollectionRemove( coll, ce.getLoadedKey() ) ) {
+                                       actionQueue.addAction(
+                                                       new CollectionRemoveAction(
+                                                                       coll,
+                                                                       ce.getLoadedPersister(),
+                                                                       ce.getLoadedKey(),
+                                                                       ce.isSnapshotEmpty(coll),
+                                                                       session
+                                                               )
+                                               );
+                               }
+
 			}
 			if ( ce.isDoupdate() ) {
-				session.getInterceptor().onCollectionUpdate( coll, ce.getLoadedKey() );
-				actionQueue.addAction(
+				if ( session.getInterceptor().onCollectionUpdate( coll, ce.getLoadedKey() ) ) {
+					actionQueue.addAction(
 						new CollectionUpdateAction(
 								coll,
 								ce.getLoadedPersister(),
@@ -290,6 +315,7 @@ public abstract class AbstractFlushingEventListener implements Serializable {
 								session
 							)
 					);
+				}
 			}
 
 		}
-- 
1.7.4.4

